Goal:	When newgrp process sits between parent and child shells, it should
	propagate STOPs from child to parent and CONTs from parent to child,
	otherwise e.g. bash's "suspend" command won't work.

Status wrt upstream: Will be in 4.0.13

Notes:
	newgrp doesn't fork and create child process, it just calls exec(),
	so there won't be "newgrp" process between parent and child shells.

	Closes: 314727
	(suspend command from su shell fails to return to parent shell)

Index: shadow-4.0.12/src/newgrp.c
===================================================================
--- shadow-4.0.12.orig/src/newgrp.c	2005-08-17 11:51:09.000000000 +0200
+++ shadow-4.0.12/src/newgrp.c	2005-08-17 11:51:13.000000000 +0200
@@ -472,9 +472,18 @@
 			exit (1);
 		} else if (child) {
 			/* parent - wait for child to finish, then log session close */
+			int cst = 0;
 			do {
-				pid = waitpid (child, NULL, 0);
-			} while (pid != child);
+				errno = 0;
+				pid = waitpid (child, &cst, WUNTRACED);
+				if (pid == child && WIFSTOPPED(cst)) {
+					/* stop when child stops */
+					raise(SIGSTOP);
+					/* wake child when resumed */
+					kill(child, SIGCONT);
+				}
+			} while (pid == child && WIFSTOPPED(cst) ||
+					pid != child && errno == EINTR);
 			SYSLOG ((LOG_INFO,
 				 "user `%s' (login `%s' on %s) returned to group `%s'",
 				 name, loginname, tty,
